homeassistant/components/smartthings/lock.py

```diff
@@ -0,0 +1,54 @@
+"""
+Support for locks through the SmartThings cloud API.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/smartthings.lock/
+"""
+from homeassistant.components.lock import LockDevice
+from homeassistant.const import STATE_LOCKED
+
+from . import SmartThingsEntity
+from .const import DATA_BROKERS, DOMAIN
+
+DEPENDENCIES = ['smartthings']
+
+
+async def async_setup_platform(hass, config, async_add_entities,
+                               discovery_info=None):
+    """Platform uses config entry setup."""
+    pass
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Add locks for a config entry."""
+    broker = hass.data[DOMAIN][DATA_BROKERS][config_entry.entry_id]
+    async_add_entities(
+        [SmartThingsLock(device) for device in broker.devices.values()
+         if is_lock(device)])
+
+
+def is_lock(device):
+    """Determine if the device supports the lock capability."""
+    from pysmartthings import Capability
+    if Capability.lock in device.capabilities:
+        return True
+    return False
+
+
+class SmartThingsLock(SmartThingsEntity, LockDevice):
```

Comment:
Let's also add a device_state_attributes property and populate the extra attributes from the data payload (`method`, `codeId`, `timeout`).  See https://smartthings.developer.samsung.com/docs/api-ref/capabilities.html#Lock for more details.  You can get these from dictionary in `self._device.status.attributes[Attribute.lock].data`.  Make sure you check for null/missing keys as not all device-type handlers may implement this.

Reply:
Ahhh, it looks like accessing `attributes[Attribute.lock].data` won't tell us much. The items we're interested in are all top-level attributes:

```
{ 
    'checkInterval': status(value=7320, unit='s', data={'protocol': 'zigbee', 'hubHardwareId': '0000', 'offlinePingable': '1'}),
    'healthStatus': status(value=None, unit=None, data={}),
    'DeviceWatch-DeviceStatus': status(value=None, unit=None, data={}),
    'lock': status(value='locked', unit=None, data=None),
    'battery': status(value=100, unit='%', data=None),
    'codeLength': status(value=None, unit=None, data=None),
    'maxCodes': status(value=250, unit=None, data=None),
    'maxCodeLength': status(value=8, unit=None, data=None),
    'codeChanged': status(value='3 changed', unit=None, data=None),
    'minCodeLength': status(value=4, unit=None, data=None),
    'scanCodes': status(value='Complete', unit=None, data=None),
    'codeReport': status(value=[3], unit=None, data=None),
    'lockCodes': status(value='{"1":"User_1","2":"User_2","3":"User_3"}', unit=None, data=None)
}
```

I've put something in to loop through all available attributes but not sure if that's in line with how you have designed things. Would I be correct in saying that we would want to add these attributes to `pysmartthings` and assign them explicitly?

My `attributes` output showing in HA:

```
check_interval: 7320
lock: locked
battery: 100
max_codes: 250
max_code_length: 8
code_changed: 3 changed
min_code_length: 4
scan_codes: Complete
code_report: 3
lock_codes: {"1":"User_1","2":"User_2","3":"User_3"}
friendly_name: My Door Lock
```

Reply:
Your device handler isn't populating data for `lock`, but others do... and we just want the lock data, not the other attributes, as many of those will be represented by other platforms or are junk (i.e. health).  

Something like this is fine:
```python
    @property
    def device_state_attributes(self):
        """Return device specific state attributes."""
        from pysmartthings import Attribute

        data = self._device.status.attributes[Attribute.lock].data or {}
        return {
            'method': data.get('method'),
            'code_id': data.get('codeId'),
            'timeout': data.get('timeout')
        }
```

Reply:
Understood. I'm using the standard SmartThings Zigbee DTH so weird that it does things differently. Z-Wave looks similar but Danalock looks different again. 🤔

Reply:
Welcome to SmartThings.  :)

I think we should also add the raw lock state here too, maybe as ‘lock_state’ because this entity represents it as a Boolean but there are actually other states, like ‘unlocked with timeout’

Reply:
Oops, sorry I deleted my comment as I think i had completely misunderstood the way the attributes get updated 😄 
Adding the raw lock state is a good idea, I've added this in the latest commit.